File: 树/到其他节点和.go
package main

func sumOfDistancesInTree(n int, edges [][]int) []int {
	grapth := make([][]int, n)

	for _, v := range edges {
		x, y := v[0], v[1]
		grapth[x] = append(grapth[x], y)
		grapth[y] = append(grapth[y], x)
	}

	ans := make([]int, n)
	sz := make([]int, n)

	var dfs func(x, p, d int)
	dfs = func(x, p, d int) {
		ans[0] += d
		sz[x] = 1
		for _, v := range grapth[x] {
			if v != p {
				// 递归算法中，顺序看是需要从前往后算还是从后向前
				dfs(v, x, d+1)
				// 此处计算树的子节点数量，明显是从后向前算，需要先递归到头，在执行计算，所以先写递归，然后想加
				sz[x] += sz[v]
			}
		}
	}
	dfs(0, -1, 0)

	var root func(int, int)
	root = func(x, p int) {
		for _, v := range grapth[x] {
			if v != p {
				// 此处从上向下计算，动态规划每个节点的值，所以在递归之前。
				ans[v] = ans[x] + n - 2*sz[v]
				root(v, x)
			}
		}
	}
	root(0, -1)
	return ans
}

File: 树/到其他节点和.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
    vector<int> sumOfDistancesInTree(int n, vector<vector<int>>& edges) {
        vector<vector<int>> tree(n);
        vector<int> ans(n);

        // make tree
        for (auto &a : edges) {
            int x = a[0], y = a[1];
            tree[x].push_back(y);
            tree[y].push_back(x);
        }

        vector<int> sz(n, 1);
        function<void(int,int,int)> dfs = [&](int x, int father, int depth) {
            ans[0] += depth;
            for (int j : tree[x]) {
                if (j != father) {
                    dfs(j,x,depth+1);
                    sz[x] += sz[j];
                }
            }
        };
        dfs(0,-1,0);

        function<void(int, int)> root = [&](int x, int fa) {
            for (int y : tree[x]) {
                if (y!= fa) {
                    ans[y] = ans[x] + n - 2* sz[y];
                    root(y, x);
                }
            }
        };
        root(0,-1);
        return ans;

    }
};